core/arch/arm/plat-rcar/romapi.c

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, EPAM Systems
+ */
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+
+#include "rcar.h"
+#include "romapi.h"
+
+static int get_api_table_index(void)
+{
+	/*
+	 * Depending on SoC type and version, there are 4 possible addresses
+	 * for each ROMAPI function
+	 */
+	static int index __nex_data = -1;
+
+	if (index != -1)
+		return index;
+
+	switch (rcar_prr_value & PRR_PRODUCT_MASK) {
+	case PRR_PRODUCT_H3:
+		switch (rcar_prr_value & PRR_CUT_MASK) {
+		case PRR_CUT_10:	/* H3 ES1.0 */
+		case PRR_CUT_11:	/* H3 ES1.1 */
+			index = 0;
+			break;
+		case PRR_CUT_20:	/* H3 ES2.0 */
+			index = 1;
+			break;
+		default:	/* Newer H3 versions use unified table */
+			index = 3;
+			break;
+		}
+		break;
+	case PRR_PRODUCT_M3W:
+		switch (rcar_prr_value & PRR_CUT_MASK) {
+		case PRR_CUT_10:	/* M3 ES1.0 */
+			index = 2;
+			break;
+		default:	/* Newer M3 versions use unified table */
+			index = 3;
+			break;
+		}
+		break;
+	default:			/* All other SoCs use unified table */
+		index = 3;
+		break;
+	}
+
+	return index;
+}
+
+/* implemented in romapi_call.S */
+extern uint32_t __plat_call_romapi(paddr_t sp, paddr_t func, uint64_t arg1,
+				   uint64_t arg2, uint64_t arg3);
+
+static uint32_t plat_call_romapi(paddr_t func, uint64_t arg1,
+				 uint64_t arg2, uint64_t arg3)
+{
+	/*
+	 * If MMU is enabled, we need to use trampoline function that will
+	 * disable MMU and switch stack pointer to physical address. On other
+	 * hand, if MMU is disabled, we can call the ROM function directly.
+	 */
+	if (cpu_mmu_enabled()) {
+		uint32_t (*fptr)(paddr_t func, uint64_t arg1, uint64_t arg2,
+				 uint64_t arg3);
+
+		/*
+		 * With ASLR enabled __plat_call_romapi() function will be
+		 * mapped at two addresses: at random address (with the rest of
+		 * OP-TEE) and at identity address. We need to map it at
+		 * identity address and call it at identity address because this
+		 * function turns off MMU to perform ROM API call. But
+		 * __plat_call_romapi *symbol* will be relocated by ASLR
+		 * code. To get identity address of the function we need to use
+		 * virt_to_phys().
+		 */
+		fptr = (void *)virt_to_phys(__plat_call_romapi);
+
+		return fptr(func, arg1, arg2, arg3);
+	} else {
+		uint32_t (*fptr)(uint64_t arg1, uint64_t arg2, uint64_t arg3);
```

Comment:
I didn't notice this before, but it seems like the supplied function pointer `func` is called using different argument depending on if mmu is enabled of not.

Reply:
Yes. If MMU is disabled, we are jumping straight to a ROM code there. If MMU is enabled, we using the trampoline is `.S` file. Trampoline needs ROM code address, so there is one extra parameter.